src/mpl/include/mpl_atomic_by_lock.h

```diff
@@ -13,15 +13,22 @@
 
 /* defined in mpl_atomic.c */
 extern pthread_mutex_t MPL_emulation_lock;
-
-#define MPL_ATOMIC_IPC_SINGLE_CS_ENTER()         \
-    do {                                         \
-        pthread_mutex_lock(&MPL_emulation_lock); \
+extern MPL_emulation_ipl_t *MPL_emulation_interprocess_lock;
+
+#define MPL_ATOMIC_IPC_SINGLE_CS_ENTER()                                       \
+    do {                                                                       \
+        pthread_mutex_lock(&MPL_emulation_lock);                               \
+        if (MPL_emulation_interprocess_lock) {                                 \
+            pthread_mutex_lock(MPL_emulation_interprocess_lock);               \
+        }                                                                      \
```

Comment:
Do we have to lock it twice? 

Can we statically initialize it as: 
```
MPL_emulation_ipl_t *MPL_emulation_interprocess_lock = &MPL_emulation_lock;
```
Upon `MPL_atomic_interprocess_lock_init`, we'll simply replace the pointer.

Reply:
This double lock is needed to use atomic operations to handle corner cases.
If we only use a single lock (either the static mutex or this interprocess lock), atomicity gets broken when we assign and free this interprocess lock.
For example, one thread takes a statically allocated mutex while another acquires the new interprocess lock. This can cause data race.

I understand that in most cases no other thread executes atomic operations when this mutex is assigned or freed (since they happen on initialization and finalization), but it *might* happen.

```c
main_thread:
  MPI_Init_thread(..);

sub_thread:
  int flag;
  while (!flag) {
    MPI_Initialized(&flag);
    //Any function that can be called before/after MPI_Init and _Finalize.
  }
```

OPA implicitly assumes no atomic operation, for example, after finalization to deal with this issue, which just makes debugging complicated since it happens only when the emulation lock is enabled and the behavior is nondeterministic. I believe this emulation lock is supported mostly for the correctness purpose, so this double lock is acceptable.

Reply:
It is guaranteed by MPI standard that init/finalize only happen in the main thread. The only potential concurrency is another thread polling the atomic state variable. But even without atomicity or mutex, the polling won't cause problem, right?

Reply:
In most cases, it will not cause an issue, but even a polling thread can update a variable (e.g., when a thread enters the MPI global critical section), which can call a non-thread safe function in the critical section. It is extremely hard to debug.

I also want to note that if finalization is added (on freeing shared memory), the outer lock is necessary to avoid SEGV: there is no other essy way to prevent other threads from accessing already freed mutex.

Reply:
I can assure you in MPICH you don't need worry about race condition during init/finalize. This is guaranteed by MPI standard. In fact, I would argue in any program, if there is race condition during init/finalize, the program is wrong. Note that I am not saying that init-less program can't exist.

For your segv concern, all you need make sure is the main thread (that calls finalize) won't segfault by itself.

Again, the race condition during init/finalize is the problem for applications. It is the wrong problem for a library to worry.

While I agree that as a fall-back solution, we are not too worried about performance. But nearly double the cost everywhere for a concern that almost never should happen, is not wise.

Reply:
Okay, if there is such assurance, I will make it a single lock.

Reply:
By the way, you always can add your concern as a comment so we are always aware of it as we read the code.

Reply:
I looked at the standard MPI 3.1. Section 12.4 "MPI and Threads" writes
```
Regardless of whether or not the MPI implementation is thread
compliant, MPI_INITIALIZED, MPI_FINALIZED, MPI_QUERY_THREAD,
MPI_IS_THREAD_MAIN, MPI_GET_VERSION and MPI_GET_LIBRARY_VERSION must
always be thread-safe. When a thread is executing one of these
routines, if another concurrently running thread also makes an MPI
call, the outcome will be as if the calls executed in some order.
```
so `MPI_Initialized()` and `MPI_Finalized()` can be called during `MPI_Init()` and `MPI_Finalize()`. It causes SEGV if we only use a single lock in the following part, for example:
https://github.com/pmodels/mpich/blob/master/src/mpi/init/finalized.c#L65

Apart from this case, in MPL/atomic, this emulation lock is currently the only implementation that truly and portably supports interprocess atomicity (since compiler-supported ones (or even assembly ones) do not explicitly guarantee such). I believe this implementation should be correct without any assumption that other atomics do not have so that this emulation lock can be used for testing.

---

I understand your concern about the performance. I tried to optimize the global lock by combining and reordering pthread_mutex/pthread_spinlock etc, but the single interprocess mutex (you suggested) performed best overall.

How about creating `-with-mpl-atomic-primitives=fastlock` which uses a single mutex model? This is not chosen by default, but it should be useful for Jenkins tests, for example.